### ISSUE TITLE
Failed to initialize connection to Redis with Sentinel and pingBeforeActivateConnection option is true

### ISSUE THREAD

**Issue by goldobin** [lettuce-io/lettuce-core#536]
Hi,

We encountered a problem while initializing a connection to the Redis with Sentinel using following configuration of the client (we're trying to use the client in fail fast fashion):

```java
final ClientOptions opts = ClientOptions
        .builder()
        .pingBeforeActivateConnection(true)
        .cancelCommandsOnReconnectFailure(true)
        .disconnectedBehavior(ClientOptions.DisconnectedBehavior.REJECT_COMMANDS)
        .build();
```
When `pingBeforeActivateConnection` is `false` client is working as expected, if `true` rises following exception:

```
Exception in thread "main" com.lambdaworks.redis.RedisException: Currently not connected. Commands are rejected.
	at com.lambdaworks.redis.protocol.CommandHandler.write(CommandHandler.java:300)
	at com.lambdaworks.redis.RedisChannelHandler.dispatch(RedisChannelHandler.java:133)
	at com.lambdaworks.redis.sentinel.StatefulRedisSentinelConnectionImpl.dispatch(StatefulRedisSentinelConnectionImpl.java:54)
	at com.lambdaworks.redis.sentinel.RedisSentinelAsyncCommandsImpl.dispatch(RedisSentinelAsyncCommandsImpl.java:142)
	at com.lambdaworks.redis.sentinel.RedisSentinelAsyncCommandsImpl.getMasterAddrByName(RedisSentinelAsyncCommandsImpl.java:59)
	at com.lambdaworks.redis.RedisClient.lookupRedis(RedisClient.java:928)
	at com.lambdaworks.redis.RedisClient.getSocketAddress(RedisClient.java:912)
	at com.lambdaworks.redis.RedisClient.lambda$getSocketAddressSupplier$0(RedisClient.java:884)
	at com.lambdaworks.redis.ConnectionBuilder.socketAddress(ConnectionBuilder.java:123)
	at com.lambdaworks.redis.AbstractRedisClient.initializeChannel(AbstractRedisClient.java:240)
	at com.lambdaworks.redis.RedisClient.connectStateful(RedisClient.java:508)
	at com.lambdaworks.redis.RedisClient.connectStandalone(RedisClient.java:478)
	at com.lambdaworks.redis.RedisClient.connect(RedisClient.java:386)
	at com.lambdaworks.redis.RedisClient.connect(RedisClient.java:373)
	at Test.main(Test.java:27)
```

Redis version: 3.0.9, 3.2.5
Lettuce version: 4.3.2.Final

Redis setup: 
- 1 x sentinel;
- 1 x redis master;
- 1 x redis client;
- all instances on the same VM;
- VM 1 x CPU (1 HW thread), 1 GB of RAM (so pretty low profile to reproduce it faster)

Test:
```
final ClientOptions opts = ClientOptions
        .builder()
        .pingBeforeActivateConnection(true)
        .cancelCommandsOnReconnectFailure(true)
        .disconnectedBehavior(ClientOptions.DisconnectedBehavior.REJECT_COMMANDS)
        .build();

final RedisClient client = RedisClient.create("redis-sentinel://10.122.17.182:26379?sentinelMasterId=m1");
client.setOptions(opts);

for (int i=0; i < 1000; i++) {

    if (i % 100 == 0) {
        System.out.println("Iteration " + i);
    }

    final RedisAsyncCommands<String, String> commands = client.connect().async();
    commands.set("key-" + i + new Random().nextInt(), "value-1").await(5, TimeUnit.SECONDS);
}
```

**Comment by mp911de** [comment]
Thanks for the detailed bug report. The issue is caused by a synchronization in the connection activation sequence. PING before connect signals completion before the internal components are activated.

**Comment by mp911de** [comment]
That's fixed now with Lettuce 4.3.3. Lettuce 4.4 (available as snapshot) and 5.0 are not affected by this issue.

**Comment by goldobin** [comment]
Thanks a lot for fix!
